### PR TITLE
dot/network: add dht advertising and discovery; handle BlockAnnounceHandshake [kusama support]

### LINKED ISSUE #1133: Test Connecting to Kusama
## Expected Behavior

- [ ] Configure node to connect to a Kusama bootnode directly 
- [ ] Ensure a successful connection is established (i.e. libp2p doesn't throw an error)

## Current Behavior

- Can't connect

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [ ] I am planning to submit a pull request to fix this issue myself

### PR DESCRIPTION
**Pull request by noot**
## Changes

- begin advertising and peer discovery via DHT upon network startup
- update DHT to be dual (both LAN and WAN - otherwise fails to begin advertising when its only peer is a local node) 
- update BlockAnnounceHandshake handling to begin syncing if peer has a higher block number than us
- fix BABE bug where non-auth node would panic when starting epoch 2

## Tests

```
go test ./dot/network -short
```
 
to test connecting to local kusama node and exchange `BlockAnnounceHandshake`:
```
git clone https://github.com/paritytech/polkadot
cd polkadot
git checkout v0.8.26
```

follow `Build from Source` instructions here https://github.com/paritytech/polkadot

once it's done building:
```
./target/release/polkadot -lsub-libp2p=debug --chain=kusama
```

you shuld see it print the kusama node's multiaddr: eg `2021-01-05 16:35:59.687   INFO tokio-runtime-worker sub-libp2p:🔍 Discovered new external address for our node: /ip4/38.88.101.233/tcp/30333/p2p/12D3KooWScfZYBVWr1c2qQHcB8YnXgrjEuXAyCDbRkKkAhhjBijy`

in a gossamer terminal,  copy paste this multiaddr into the `--bootnodes`: (also probably want to set network log level in config file to trace)
```
./bin/gossamer --config chain/ksmcc/config.toml  init --force
./bin/gossamer --config chain/ksmcc/config.toml --bootnodes /ip4/127.0.0.1/tcp/30333/p2p/12D3KooWScfZYBVWr1c2qQHcB8YnXgrjEuXAyCDbRkKkAhhjBijy
```

wait for everything to start, if everything goes well, you should see:
```
INFO[01-05|20:28:26] received stream                          pkg=network sub-protocol=/transactions/1 caller=service.go:317
INFO[01-05|20:28:26] received stream                          pkg=network sub-protocol=/block-announces/1 caller=service.go:317
INFO[01-05|20:28:26] received message on notifications sub-protocol pkg=network sub-protocol=/block-announces/1 message="BlockAnnounceHandshake Roles=1 BestBlockNumber=1059551 BestBlockHash=0x09a749864b68f8ee744d9fed81d49cdf3169104239af6c5f0013a860e934fdb7 GenesisHash=0xb0a8d493285c2df73290dfb7e61f870f17b41801197a149ca93654499ea3dafe" caller=notifications.go:104
```

the kusama node opened some streams with us and sent us a `BlockAnnounceHandshake`, if trace logs are on you'll also see the messages we sent. the node should not disconnect from us at this point

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [x] I have reviewed my own pull request before requesting a review
- [x] All integration tests and required coverage checks are passing

## Issues

- closes  #1133

### CODE REVIEW

**Review by RyRy79261 — APPROVED** [review]
Wasn't able to test kusama on Arch but standard test passed & code looks good

**Comment by edwardmack** [comment]
This is what I'm getting when trying to connect to kusama: 
```
INFO[01-07|10:55:47] Started listening                        pkg=network address=/ip4/192.168.11.41/tcp/7001/p2p/12D3KooWAFR1wQX3pJgaaDWmWfWXpft6CKhR3FaMPTeyd6KAUQRZ caller=service.go:196
INFO[01-07|10:55:47] Started listening                        pkg=network address=/ip4/127.0.0.1/tcp/7001/p2p/12D3KooWAFR1wQX3pJgaaDWmWfWXpft6CKhR3FaMPTeyd6KAUQRZ caller=service.go:196
INFO[01-07|10:55:47] started network service                  pkg=network supported protocols="[/p2p/id/delta/1.0.0 /ipfs/id/1.0.0 /ipfs/id/push/1.0.0 /ipfs/ping/1.0.0 /ksmcc3/lan/kad/1.0.0 /ksmcc3/sync/2 /ksmcc3/light/2 /ksmcc3/block-announces/1 /ksmcc3/transactions/1]" caller=service.go:219
INFO[01-07|10:55:47] registered notifications sub-protocol    pkg=network sub-protocol=/paritytech/grandpa/1 caller=service.go:334
INFO[01-07|10:55:47] current epoch                            pkg=babe epoch=1 slots into epoch=0 caller=babe.go:390
EROR[01-07|10:55:48] failed to begin DHT discovery            pkg=network error="failed to begin advertising: failed to find any peer in table" caller=service.go:214
EROR[01-07|10:56:02] Failed to read LEB128 encoding           pkg=network protocol=/ksmcc3/block-announces/1 error="stream reset" caller=service.go:460
EROR[01-07|10:56:02] Failed to read LEB128 encoding           pkg=network protocol=/ksmcc3/transactions/1 error="stream reset" caller=service.go:460

```

File: dot/network/block_announce.go [hunk 1]
```diff
@@ -103,7 +103,7 @@ func decodeBlockAnnounceMessage(r io.Reader) (NotificationsMessage, error) {
 // BlockAnnounceHandshake is exchanged by nodes that are beginning the BlockAnnounce protocol
 type BlockAnnounceHandshake struct {
 	Roles           byte
-	BestBlockNumber uint64
+	BestBlockNumber uint32
```
Comment (edwardmack) [inline dot/network/block_announce.go:106]: Why did this change from uint64 to uint32?

**Review by edwardmack — APPROVED** [review]
I wasn't able to confirm kusama connection on my local setup, however unit tests pass and code looks good.

**Comment by noot** [comment]
> This is what I'm getting when trying to connect to kusama:
> 
> ```
> INFO[01-07|10:55:47] Started listening                        pkg=network address=/ip4/192.168.11.41/tcp/7001/p2p/12D3KooWAFR1wQX3pJgaaDWmWfWXpft6CKhR3FaMPTeyd6KAUQRZ caller=service.go:196
> INFO[01-07|10:55:47] Started listening                        pkg=network address=/ip4/127.0.0.1/tcp/7001/p2p/12D3KooWAFR1wQX3pJgaaDWmWfWXpft6CKhR3FaMPTeyd6KAUQRZ caller=service.go:196
> INFO[01-07|10:55:47] started network service                  pkg=network supported protocols="[/p2p/id/delta/1.0.0 /ipfs/id/1.0.0 /ipfs/id/push/1.0.0 /ipfs/ping/1.0.0 /ksmcc3/lan/kad/1.0.0 /ksmcc3/sync/2 /ksmcc3/light/2 /ksmcc3/block-announces/1 /ksmcc3/transactions/1]" caller=service.go:219
> INFO[01-07|10:55:47] registered notifications sub-protocol    pkg=network sub-protocol=/paritytech/grandpa/1 caller=service.go:334
> INFO[01-07|10:55:47] current epoch                            pkg=babe epoch=1 slots into epoch=0 caller=babe.go:390
> EROR[01-07|10:55:48] failed to begin DHT discovery            pkg=network error="failed to begin advertising: failed to find any peer in table" caller=service.go:214
> EROR[01-07|10:56:02] Failed to read LEB128 encoding           pkg=network protocol=/ksmcc3/block-announces/1 error="stream reset" caller=service.go:460
> EROR[01-07|10:56:02] Failed to read LEB128 encoding           pkg=network protocol=/ksmcc3/transactions/1 error="stream reset" caller=service.go:460
> ```

can you put the network into trace mode and see what happens? are you using a local kusama node?

File: dot/network/block_announce.go (hunk 1 above)
Comment (noot) [reply dot/network/block_announce.go:106]: when I got the message from the kusama node it was 4 bytes for this field, not 8, will confirm with spec team

**Pull request merged**